### PR TITLE
[1.17] Add Entity Visibility API

### LINKED ISSUE #5973: Add Entity Visibility API
### Is your feature request related to a problem?

Right now, per-player entities are done mostly with sending packets. This is a hacky and unreliable way and requires either NMS or ProtocolLib usage and is prone to break with minecraft updates.

### Describe the solution you'd like.

I'd like Paper to expand upon the `Player#hidePlayer` and `Player#showPlayer` methods with new methods `Player#showEntity(Entity)`, `Player#hideEntity(Entity)` and `Player#canSee(Entity)` as well as overloading `World#spawn` methods with a boolean `shownByDefault` and adding method `Entity#setShownByDefault` and `Entity#isShownByDefault`.

### Describe alternatives you've considered.

none

### Agreements

- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have ensured the feature I'm requesting isn't already in the latest supported Paper build.

### Other

_No response_

### PR DESCRIPTION
**Pull request by Janmm14**
Closes #5973 

This is a port to 1.17 of #5993

### CODE REVIEW

**Review by Trigary — CHANGES_REQUESTED** [review]
Unfortunately `Level#sendParticles` also exists and is used by entities (eg. `Squid#spawnInk`, `LivingEntity#checkFallDamage`). You would need to handle that similarly to sounds.

**Comment by Janmm14** [comment]
> 
> 
> Unfortunately `Level#sendParticles` also exists and is used by entities (eg. `Squid#spawnInk`, `LivingEntity#checkFallDamage`). You would need to handle that similarly to sounds.

I've now edited sendParticle and playSound for all entities

**Comment by Janmm14** [comment]
stuff fixed

**Comment by Janmm14** [comment]
bump

**Comment by Janmm14** [comment]
.

File: patches/api/0321-Add-Entity-Visibility-API.patch [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     Set<Player> getTrackedPlayers();
+     // Paper end
+ 
++    // Paper start - Entity Visibility API
++    /**
++     * Checks whether this player can currently see the given entity
++     * @param entity the entity to check
++     * @return True if this player can see the given entity
++     */
++    boolean canSee(@NotNull org.bukkit.entity.Entity entity);
```
Comment (MiniDigger) [inline patches/api/0321-Add-Entity-Visibility-API.patch]: is the FQN required here? the class is in the same package

File: patches/server/0719-Add-Entity-Visibility-API.patch [hunk 2]
```diff
@@ -0,0 +1,748 @@
...
+diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
+index 7f7bc04a30a0422b2d589adb488082c0aa5326dc..7c32f439078b787bd28407d479530b445e49af6d 100644
+--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
++++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
+@@ -2277,7 +2277,7 @@ Sections go from 0..16. Now whenever a section is not empty, it can potentially
+         // Paper start - use distance map to optimise tracker
+         com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> lastTrackerCandidates;
+ 
+-        final void updatePlayers(com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> newTrackerCandidates) {
++        public final void updatePlayers(com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> newTrackerCandidates) {
```
Comment (MiniDigger) [inline patches/server/0719-Add-Entity-Visibility-API.patch]: this is missing a paper comment, and really, it should be a AT

File: patches/server/0719-Add-Entity-Visibility-API.patch [hunk 3]
```diff
@@ -0,0 +1,748 @@
...
+diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
+index ac5779319081a6894373877067edf958da8a9cf5..9ca996841989f377d29250e323ce2ecc06f5a638 100644
+--- a/src/main/java/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
++++ b/src/main/java/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
+@@ -28,7 +28,7 @@ import org.bukkit.event.entity.EntityInteractEvent;
+ public class RemoveBlockGoal extends MoveToBlockGoal {
+ 
+     private final Block blockToRemove;
+-    private final Mob removerMob;
++    public final Mob removerMob; // Paper - private -> public - Entity Visibility API
```
Comment (MiniDigger) [inline patches/server/0719-Add-Entity-Visibility-API.patch]: AT

File: patches/api/0321-Add-Entity-Visibility-API.patch [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
+index a9e455c5b3bbe4edbdb71f86f5c6eebc2f605547..8e7f0129f58a1d2bcb6f6f521637845da22b1be1 100644
+--- a/src/main/java/org/bukkit/entity/Entity.java
++++ b/src/main/java/org/bukkit/entity/Entity.java
+@@ -753,4 +753,18 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
+      */
+     public boolean isTicking();
+     // Paper end
++
++    // Paper start - Entity Visibility API
```
Comment (NoahvdAa) [inline patches/api/0321-Add-Entity-Visibility-API.patch:16]: Don't create a new comment "block" here, move it into the already existing one

File: patches/api/0321-Add-Entity-Visibility-API.patch [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ }
+diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
+index da83b4cbed0be6f693c7cbb1cc032356f12d7883..71b6fe4a92d2d914bc4436513abd210c0fd7bf6c 100644
+--- a/src/main/java/org/bukkit/entity/Player.java
++++ b/src/main/java/org/bukkit/entity/Player.java
+@@ -1986,6 +1986,27 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
+     Set<Player> getTrackedPlayers();
+     // Paper end
+ 
++    // Paper start - Entity Visibility API
```
Comment (NoahvdAa) [inline patches/api/0321-Add-Entity-Visibility-API.patch:38]: same here

File: patches/server/0719-Add-Entity-Visibility-API.patch [hunk 6]
```diff
@@ -0,0 +1,748 @@
...
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+index 8246ad7ebecdfc0b7519fe4412fef7b07407e850..40eb71c060c11d95e3c4962712d178a83fe2212e 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+@@ -1207,4 +1207,23 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
+         return getHandle().isTicking();
+     }
+     // Paper end
++
++    // Paper start - Entity Visibility API
```
Comment (NoahvdAa) [inline patches/server/0719-Add-Entity-Visibility-API.patch:635]: Same here

File: patches/server/0719-Add-Entity-Visibility-API.patch [hunk 7]
```diff
@@ -0,0 +1,748 @@
...
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -2473,4 +2473,63 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         return this.spigot;
+     }
+     // Spigot end
++
++    // Paper start - Entity Visibility API
++    @Override
++    public boolean canSee(org.bukkit.entity.Entity entity) {
```
Comment (NoahvdAa) [inline patches/server/0719-Add-Entity-Visibility-API.patch:665]: Would it be a good idea to null check the entity? (same for methods below)

File: patches/server/0719-Add-Entity-Visibility-API.patch (hunk 2 above)
Comment (Janmm14) [reply patches/server/0719-Add-Entity-Visibility-API.patch]: "AT"?

Comment (lynxplay) [reply patches/server/0719-Add-Entity-Visibility-API.patch]: https://github.com/PaperMC/Paper/blob/master/build-data/paper.at allows you to change access modifiers.

**Comment by jaylawl** [comment]
Please don't let this die

File: patches/api/0321-Add-Entity-Visibility-API.patch (hunk 4 above)
Comment (Janmm14) [reply patches/api/0321-Add-Entity-Visibility-API.patch:16]: But its done like that in many more places and helps with keeping patches individual.

File: patches/api/0321-Add-Entity-Visibility-API.patch (hunk 5 above)
Comment (Janmm14) [reply patches/api/0321-Add-Entity-Visibility-API.patch:38]: But litterally a couple lines above the same thing happens already and I think that it also helps to keep patches a little more individual.

File: patches/server/0719-Add-Entity-Visibility-API.patch (hunk 2 above)
Comment (Janmm14) [reply patches/server/0719-Add-Entity-Visibility-API.patch]: Cannot use AT for this one because method is added by patch 0431

Comment (Machine-Maker) [reply patches/server/0719-Add-Entity-Visibility-API.patch]: then go back to the patch that adds it and change the visibility there.

File: patches/server/0719-Add-Entity-Visibility-API.patch (hunk 3 above)
Comment (Janmm14) [reply patches/server/0719-Add-Entity-Visibility-API.patch]: done

File: patches/api/0321-Add-Entity-Visibility-API.patch (hunk 1 above)
Comment (Janmm14) [reply patches/api/0321-Add-Entity-Visibility-API.patch]: fixed

File: patches/server/0719-Add-Entity-Visibility-API.patch (hunk 7 above)
Comment (Janmm14) [reply patches/server/0719-Add-Entity-Visibility-API.patch:665]: null checks added

File: patches/server/0719-Add-Entity-Visibility-API.patch [hunk 8]
```diff
@@ -0,0 +1,742 @@
...
+diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
+index 7f7bc04a30a0422b2d589adb488082c0aa5326dc..b72ec569617a2b714bb5adcbba884ca7d6f1405f 100644
+--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
++++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
+@@ -2277,7 +2277,7 @@ Sections go from 0..16. Now whenever a section is not empty, it can potentially
+         // Paper start - use distance map to optimise tracker
+         com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> lastTrackerCandidates;
+ 
+-        final void updatePlayers(com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> newTrackerCandidates) {
++        public final void updatePlayers(com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> newTrackerCandidates) { // Paper - Entity Visibility API - packageprivate -> public
```
Comment (Machine-Maker) [inline patches/server/0719-Add-Entity-Visibility-API.patch:16]: Unsure if you saw my comment [here](https://github.com/PaperMC/Paper/pull/6000#discussion_r758742147), but if this is a method added by paper in a previous patch, you should go back to that patch and change the visibility there to reduce the diff.

**Comment by Janmm14** [comment]
soo I remembered again I had this thing open, please check again

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
Just as a note, [upstream added something](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/commits/37fd1917e73f1f0b16dcef01cac940cd52510d73) that might conflict with this. Probably needs to be considered somehow here.

**Comment by Janmm14** [comment]
> @Machine-Maker wrote:
>
> Just as a note, [upstream added something](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/commits/37fd1917e73f1f0b16dcef01cac940cd52510d73) that might conflict with this. Probably needs to be considered somehow here.

Looks like this patch got a little redundant.

I will create a PR to upstream with the details upstream ignored (sounds, particles).
As they accepted the base idea, the fixes to it will hopefully be accepted as well.

Edit: So far I did not create any upstream PR for fixes like sound and I don't plan to in the near future, so if anyone wants to do it instead, go ahead. If not, also fine.

**Comment by Owen1212055** [comment]
Seeing as upstream has merged _most_ of what this pr was meant to bring this will be closed.

The whole "hide by default behavior" seems to be covered by another issue which has some PRs for it as well. https://github.com/PaperMC/Paper/issues/7395

Because blocking sounds/particles might still be a wanted feature, feel free to open a separate PR (here or upstream) that covers that specific area while using the current visibility API that upstream has.

Thank you for your contribution, and sorry we couldn't get to merge this. ☹️

**Pull request closed without merging**